### ISSUE TITLE
Private channel buffer is empty

### ISSUE THREAD

**Issue by liskin** [wee-slack/wee-slack#664]
Apparently Slack now has private channels that aren't groups but don't work with the channels API anyway. These are listed among `"channels": […]` and have `"is_private": true`.

I tried to fix this myself: https://github.com/liskin/wee-slack/commits/fix-private-channels
Unfortunately the `conversations.mark` API method isn't documented and doesn't work with the legacy token, only with a token stolen from the web client. There's a twitter thread with `@SlackHQ` about it: https://twitter.com/Saythlin/status/1046131572808863745 without a solution at the time of writing.

**Comment by adamgibbins** [comment]
When you create a new channel in the web client or native apps theres a slider at the top to make it public or private.

**Comment by trygveaa** [comment]
That creates a group for me, not what @liskin describes.

**Comment by liskin** [comment]
I have no idea how my colleagues managed to create one. I suspect they created a normal channel and later made it private, or something like that.

**Comment by liskin** [comment]
Oh and in the web client they show up as groups.

**Comment by trygveaa** [comment]
Ah, you're right. Converting a public channel to a private one did it. So I can reproduce the problem at least. But I guess we'll have to wait and see what Slack responds in that twitter thread.

**Comment by liskin** [comment]
Yeah, in the meantime I found out that with the conversations API, I don't see old threads (but I do see _some_ threads), and I probably need to use `conversations.replies` to get old thread histories. Bah. :-/

**Comment by liskin** [comment]
Their just replied to me via e-mail:

> Thanks for reaching out about marking converted private channels as read/marked.  Apologies for the troubles here.
>
> `conversations.mark` is a hidden method that has not been released for public use via our Conversations API.  This is the reason you see
the token type error messages as it is not accepting the `xoxp` token associated with your app.
>
> I know this has come up before from other users who need to use the Conversations API for dealing with shared channels, for example.  I
found a few discussions about this but so far it has not been decided if this method will become available or not.
>
> I can share your intentions with this method as you have identified that some private channels will not respond to `groups.mark`.  In
the meantime, apologies for the limitation.
>
> Let me know if you have any related thoughts/questions.

**Comment by trygveaa** [comment]
:/
So we'll just have to wait and see I guess.

We can fix history and such in the meantime though.

**Comment by liskin** [comment]
Well, `conversations.mark` works for me, just not with the legacy token so I'm not creating a pull request.

**Comment by trygveaa** [comment]
Sure, but history loading can be fixed independently of marking.

**Comment by liskin** [comment]
Without breaking threads, I'm afraid we can't use any of my fixes. I'm trying to make it work somehow, but it'll take a while. :-(

**Comment by trygveaa** [comment]
I have fixed some of this in https://github.com/trygveaa/wee-slack/tree/private-channels, but it's not complete yet.

The things which are not working yet:
- `thread_messages_in_channel` is only working for messages which arrive after the channel history has loaded.
- If `background_load_all_history` is false, a private channel will not show as unread in the hotlist when wee-slack loads, even if there are unread messages. Messages which arrive after wee-slack has loaded will still mark the channel as unread in the hotlist.
- Which messages that has been read is not synced to Slack.

The last issue is not possible to fix without API changes from Slack. I'm not sure if the second issue is possible to fix without API changes either.

There may be more issues, but these are the ones I'm aware of. Note that these issues only exist for private channels (i.e. a channel which was public, but has been converted to private). Other channels should still work fine without these issues.

**Comment by liskin** [comment]
Here's another reply from Slack support:

> Most of the work surrounding our API is directed towards the Conversations API which is meant to address inconsistencies and troubles
> while juggling both `groups.*` and `channels.*` methods or scopes.  That said, there are instances where there is not feature parity
> (`conversations.mark`) or differences such as how replies are structured.
>
> I know that it does take an extra call to `conversations.replies` to fetch replies for a parent message but I don't believe this is
> something that will be adjusted.  But I will make the case for `conversations.mark` as I think it is important to have parity when
> asking developers to switch to the Conversations API.

So I think once they make `conversations.mark` available, we should switch to the new API (which unfortunately means a separate API call for listing users in channels, and maybe something more about the unread count or something). I've been busy the last week so I haven't tried your branch yet, but it seems it'd work, including marking channels read if webclient token is used. I'll give it a go later this week. Threads are still a bit buggy, e.g. the `[ Thread: ... Replies: … ]` marker relies on actually having those replies, whereas the conversations API only gives the count, making old threads invisible. That'll need to be fixed.

**Comment by liskin** [comment]
Oh and also, we're completely ignoring thread broadcasts (replies that should be visible in channel).

**Comment by mkaito** [comment]
Can confirm this happens. One of the channels I interact with the most used to be public (mistake when creating it) and was made private. Now wee-slack can't load history or nicks for it. And given that wee-slack seems to reload channel history nearly every time you switch to a buffer now, I can pretty much not use this channel at all.

**Comment by hydrobiont** [comment]
> Can confirm this happens. One of the channels I interact with the most used to be public (mistake when creating it) and was made private. Now wee-slack can't load history or nicks for it. And given that wee-slack seems to reload channel history nearly every time you switch to a buffer now, I can pretty much not use this channel at all.

Also can confirm exactly the same behavior

**Comment by kthibodeaux** [comment]
> Can confirm this happens. One of the channels I interact with the most used to be public (mistake when creating it) and was made private. Now wee-slack can't load history or nicks for it. And given that wee-slack seems to reload channel history nearly every time you switch to a buffer now, I can pretty much not use this channel at all.

This is happening for me, too.

**Comment by carwin** [comment]
Has there been any more news from the Slack team about this?

**Comment by liskin** [comment]
Not really, but even if there was, I no longer have the company e-mail that was used to communicate with them. (And, therefore, I no longer have this problem. :-))

**Comment by jcross** [comment]
Could we get at least get the partial fix from @trygveaa merged in? The caveats of his fix are better than the current situation.

**Comment by trygveaa** [comment]
@jcross: Yeah, I merged it now.

So now, these channels should work except for [the limitations documented in the readme](https://github.com/wee-slack/wee-slack#known-issues) (the same as the ones I wrote in my last comment above).

**Comment by trygveaa** [comment]
I think I've done as much as I can to support this now. This is the current state of the remaining issues [I mentioned previously](https://github.com/wee-slack/wee-slack/issues/664#issuecomment-452482539):

> thread_messages_in_channel is only working for messages which arrive after the channel history has loaded.

This has been fixed.

> If background_load_all_history is false, a private channel will not show as unread in the hotlist when wee-slack loads, even if there are unread messages. Messages which arrive after wee-slack has loaded will still mark the channel as unread in the hotlist.

I asked Slack about this, but they don't have any API to check for unread messages apart from loading the history, so I've [noted this in the readme](https://github.com/wee-slack/wee-slack#known-issues). If you have `background_load_all_history` set to true (which is now the default), it works fine.

> Which messages that has been read is not synced to Slack.

I asked Slack about this too, but they don't want to support this. It is possible to do when using a session token though, so I've added support for that. I.e. it will not work when using a legacy token or an OAuth token, but it will work when using a session token. I've described [the differences between the tokens in the readme](https://github.com/wee-slack/wee-slack#4-add-your-slack-api-tokens). I've also noted under [known issues](https://github.com/wee-slack/wee-slack#known-issues) that it's not working with legacy/OAuth tokens.

Also note that Slack is going to remove the API we use for normal channels in November, so we have to switch to the same API as we use for these private channels. This means that all channels will be handled the same, and the above limitations will be for all channels.
